src/lib/logical_query_plan/abstract_lqp_node.cpp

```diff
@@ -169,86 +189,139 @@ const std::vector<std::string>& AbstractLQPNode::output_column_names() const {
+    for (auto column_id = ColumnID{0}; column_id < output_column_count(); ++column_id) {
+      if (output_column_names()[column_id] == qualified_column_name_without_local_table_name->column_name) {
+        return output_column_references()[column_id];
+      }
+    }
+    return std::nullopt;
   }
-}
 
-bool AbstractLQPNode::knows_table(const std::string& table_name) const {
   /**
-   * This function might have to be overwritten if a node can handle different input tables, e.g. a JOIN.
+   * Look for the Column in child nodes
    */
-  DebugAssert(left_child() && !right_child(),
-              "Node has no or two inputs and therefore needs to override this function");
-  if (_table_alias) {
-    return *_table_alias == table_name;
-  } else {
-    return left_child()->knows_table(table_name);
+  const auto resolve_qualified_column_name = [&](
+      const auto& node, const auto& qualified_column_name) -> std::optional<LQPColumnReference> {
+    if (node) {
+      const auto column_reference = node->find_column_reference(qualified_column_name);
+      if (column_reference) {
+        if (find_output_column_id_by_column_reference(*column_reference)) {
+          return column_reference;
+        }
+      }
+    }
+    return std::nullopt;
+  };
+
+  const auto column_reference_from_left =
+      resolve_qualified_column_name(left_child(), *qualified_column_name_without_local_table_name);
+  const auto column_reference_from_right =
+      resolve_qualified_column_name(right_child(), *qualified_column_name_without_local_table_name);
+
+  Assert(!column_reference_from_left || !column_reference_from_right ||
+             column_reference_from_left == column_reference_from_right,
+         "Column '" + qualified_column_name_without_local_table_name->as_string() + "' is ambiguous");
+
+  if (column_reference_from_left) {
+    return column_reference_from_left;
   }
+  return column_reference_from_right;
 }
 
-std::vector<ColumnID> AbstractLQPNode::get_output_column_ids() const {
-  std::vector<ColumnID> column_ids(output_column_count());
-  std::iota(column_ids.begin(), column_ids.end(), 0);
-  return column_ids;
+LQPColumnReference AbstractLQPNode::get_column_reference(const QualifiedColumnName& qualified_column_name) const {
+  const auto colum_origin = find_column_reference(qualified_column_name);
+  DebugAssert(colum_origin, "Couldn't resolve column origin of " + qualified_column_name.as_string());
+  return *colum_origin;
 }
 
-std::vector<ColumnID> AbstractLQPNode::get_output_column_ids_for_table(const std::string& table_name) const {
-  /**
-   * This function might have to be overwritten if a node can handle different input tables, e.g. a JOIN.
-   */
-  DebugAssert(left_child() && !right_child(),
-              "Node has no or two inputs and therefore needs to override this function.");
+std::shared_ptr<const AbstractLQPNode> AbstractLQPNode::find_table_name_origin(const std::string& table_name) const {
```

Comment:
If we rename `find_column_reference` to `find_column`, we could rename this to `find_table`.

Reply:
nah, `table_name` as the word for an alias or the stored_table_name needs to stay in, imo.